crates/runtime/src/datafusion/cluster/codec/spice_logical_codec.rs

```diff
@@ -0,0 +1,111 @@
+use crate::Runtime;
+use crate::datafusion::extension::bytes_processed::BytesProcessedNode;
+use arrow_schema::SchemaRef;
+use ballista_core::serde::BallistaLogicalExtensionCodec;
+use datafusion::catalog::TableProvider;
+use datafusion::common::{DataFusionError, Result, TableReference, exec_err};
+use datafusion::prelude::SessionContext;
+use datafusion_expr::registry::FunctionRegistry;
+use datafusion_expr::{Extension, LogicalPlan, ScalarUDF};
+use datafusion_proto::logical_plan::LogicalExtensionCodec;
+use std::fmt::Debug;
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::task;
+
+/// Serialization support for custom Spice logical nodes
+pub struct SpiceLogicalCodec {
+    inner: Arc<dyn LogicalExtensionCodec>,
+    runtime: Option<Arc<Runtime>>,
+}
+
+impl Debug for SpiceLogicalCodec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SpiceLogicalCodec")
+    }
+}
+
+impl SpiceLogicalCodec {
+    #[must_use]
+    pub fn new_with_runtime(runtime: Arc<Runtime>) -> Arc<dyn LogicalExtensionCodec> {
+        Arc::new(Self {
+            inner: Arc::new(BallistaLogicalExtensionCodec::default()),
+            runtime: Some(runtime),
+        })
+    }
+
+    #[must_use]
+    pub fn new_codec() -> Arc<dyn LogicalExtensionCodec> {
+        Arc::new(Self {
+            inner: Arc::new(BallistaLogicalExtensionCodec::default()),
+            runtime: None,
+        })
+    }
+
+    fn runtime(&self) -> Result<Arc<Runtime>> {
+        self.runtime.clone().ok_or(DataFusionError::Execution(
+            "SpiceLogicalCodec did not bind a Runtime handle. This is a bug.".to_string(),
```

Comment:
```suggestion
            "SpiceLogicalCodec did not bind a Runtime handle. Report this bug on GitHub: https://github.com/spiceai/spiceai/issues".to_string(),
```

I think this will display as a datafusion error, but we probably want to direct users to submit a bug on our repo - not datafusion. This could look weird if the datafusion error is saying to submit a bug on datafusion, while we're putting a message to submit on spice as well ☹️

Reply:
Good note. Sadly re the error type, DF traits want to consume `datafusion::common::Error`.